File: Haskell/Automato-Plauground/app/Automato.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE DeriveGeneric #-}

-- Módulo Automato: Define as funções e tipos relacionados ao autômato.
module Automato
  ( Automato, -- Exporta o tipo Automato
    criarAutomato, -- Função para criar um autômato a partir de um arquivo JSON
    imprimirAutomato, -- Função para exibir informações do autômato
    testePalavra, -- Função para testar uma palavra no autômato
    salvarResultados, -- Função para salvar os resultados dos testes em JSON
    ResultadoTeste (..) -- Exporta o construtor e os campos de ResultadoTeste
  )
where

-- Importações necessárias
import Data.Aeson (FromJSON, ToJSON, decode, object, (.=)) -- Para trabalhar com JSON
import Data.Aeson.Encode.Pretty (encodePretty) -- Para gerar JSON formatado ("pretty")
import qualified Data.ByteString.Lazy as B -- Para manipular arquivos binários (JSON)
import GHC.Generics (Generic) -- Para derivar instâncias de FromJSON e ToJSON automaticamente
import Data.Set (Set, member) -- Para representar conjuntos (alfabeto e estados finais)
import qualified Data.Map.Strict as Map -- Para representar transições como um mapa

-- Definição do tipo Automato
data Automato = Automato
  { alfabeto       :: Set String -- Conjunto de símbolos do alfabeto
  , estadoInicial  :: String -- Estado inicial do autômato
  , estadosFinais  :: Set String -- Conjunto de estados finais
  , transicoes     :: Map.Map String (Map.Map String String) -- Transições do autômato
  } deriving (Show, Generic)

-- Instância para permitir que o tipo Automato seja decodificado de JSON
instance FromJSON Automato

-- Definição do tipo ResultadoTeste
data ResultadoTeste = ResultadoTeste
  { palavra          :: String -- A palavra testada
  , aceita           :: Bool -- Indica se a palavra foi aceita ou rejeitada
  , estadosPercorridos :: [String] -- Lista dos estados percorridos durante o teste
  } deriving (Show, Generic)

-- Instância para permitir que o tipo ResultadoTeste seja codificado em JSON
instance ToJSON ResultadoTeste

-- Função para ler o arquivo JSON contendo a definição do autômato
lerJson :: FilePath -> IO (Maybe Automato)
lerJson arquivo = do
  conteudo <- B.readFile arquivo -- Lê o conteúdo do arquivo JSON
  return (decode conteudo) -- Decodifica o JSON para o tipo Automato

-- Função para criar o autômato a partir do arquivo JSON
criarAutomato :: FilePath -> IO Automato
criarAutomato arquivo = do
  maybeAutomato <- lerJson arquivo -- Tenta carregar o autômato do JSON
  case maybeAutomato of
    Just automato -> return automato -- Retorna o autômato se o JSON for válido
    Nothing       -> error "Erro ao carregar o autômato do JSON" -- Exibe erro se falhar

-- Função para imprimir informações detalhadas sobre o autômato
imprimirAutomato :: Automato -> String
imprimirAutomato automato =
  "Alfabeto: " ++ show (alfabeto automato) ++ "\n" ++
  "Estado Inicial: " ++ estadoInicial automato ++ "\n" ++
  "Estado(s) Final(is): " ++ show (estadosFinais automato) ++ "\n" ++
  "Transições:\n" ++
  concatMap (\(estado, trans) ->
               "  " ++ estado ++ " -> " ++ show trans ++ "\n")
            (Map.toList (transicoes automato)) -- Converte as transições para string

-- Função para testar uma palavra no autômato
testePalavra :: Automato -> String -> IO ResultadoTeste
testePalavra automato palavraTestada =
  let simbolos = map (:[]) palavraTestada -- Converte cada caractere da palavra em uma string
      invalida = any (\s -> not (member s (alfabeto automato))) simbolos -- Verifica se a palavra está no alfabeto
  in if invalida
     then return $ ResultadoTeste palavraTestada False ["Fora do alfabeto"] -- Rejeita palavras fora do alfabeto
     else do
       -- Simula as transições do autômato
       let (estadoFinal, estadosSimulados) = foldl (\(estadoAtual, estados) letra ->
                                  case Map.lookup estadoAtual (transicoes automato) >>= Map.lookup letra of
                                    Just proximoEstado -> (proximoEstado, proximoEstado : estados) -- Avança para o próximo estado
                                    Nothing            -> ("", []) -- Trata transições inválidas
                               )
                               (estadoInicial automato, [estadoInicial automato]) -- Começa no estado inicial
                               simbolos

       -- Retorna o resultado do teste
       return $ ResultadoTeste palavraTestada (estadoFinal /= "" && member estadoFinal (estadosFinais automato)) (reverse estadosSimulados)

-- Função para salvar os resultados dos testes em um arquivo JSON
salvarResultados :: FilePath -> [ResultadoTeste] -> IO ()
salvarResultados arquivo resultados = do
  let json = encodePretty $ object ["resultados" .= resultados] -- Codifica os resultados em JSON formatado
  B.writeFile arquivo json -- Salva o JSON em um arquivo
  putStrLn $ "Resultados salvos em " ++ arquivo -- Informa ao usuário onde os resultados foram salvos

File: Prolog/automato/automato.pl
:- module(automato, [
    criar_automato/2, % Cria um autômato a partir de um arquivo JSON
    imprimir_automato/1, % Exibe as informações do autômato
    teste_palavra/3, % Verifica se uma palavra é aceita pelo autômato
    salvar_resultados/2 % Exporta resultados de testes para um arquivo JSON
]).

:- use_module(library(http/json)). % Biblioteca para leitura e escrita de JSON
:- use_module(library(http/json_convert)). % Biblioteca para conversão Prolog-JSON


% Carrega um arquivo JSON e o converte em estrutura Prolog
ler_json(Arquivo, Json) :-
    open(Arquivo, read, Stream), % Abre o arquivo em modo leitura
    json_read(Stream, Json), % Converte o conteúdo JSON em Prolog
    close(Stream). % Fecha o fluxo de leitura

% Transforma as transições do formato JSON em uma lista de tuplas
extrair_transicoes(json(TransicoesJSON), Transicoes) :-
    findall(de(EstadoOrigem, Simbolo, EstadoDestino),
            ( member(EstadoOrigem=json(Simbolos), TransicoesJSON),
              member(Simbolo=EstadoDestino, Simbolos)
            ),
            Transicoes).

% Cria um autômato a partir de um arquivo JSON
criar_automato(Arquivo, Automato) :-
    ler_json(Arquivo, Json), % Carrega o conteúdo do arquivo JSON
    Json = json([alfabeto=Alfabeto, estadoInicial=EstadoInicial, estadosFinais=EstadosFinais, transicoes=TransicoesJSON]),
    extrair_transicoes(TransicoesJSON, Transicoes), % Converte transições do JSON para lista
    Automato = automato(Alfabeto, EstadoInicial, EstadosFinais, Transicoes). % Define a estrutura do autômato

% Exibe as informações do autômato no console
imprimir_automato(automato(Alfabeto, EstadoInicial, EstadosFinais, Transicoes)) :-
    format("=== Informacoes do Automato ===~n", []), % Cabeçalho da exibição
    format("Alfabeto: ~w~n", [Alfabeto]), % Lista os símbolos do alfabeto
    format("Estado Inicial: ~w~n", [EstadoInicial]), % Mostra o estado inicial
    format("Estado(s) Final(is): ~w~n", [EstadosFinais]), % Lista os estados finais
    format("Transicoes:~n"), % Cabeçalho das transições
    maplist(imprimir_transicao, Transicoes). % Exibe cada transição

% Formata e exibe uma transição no console
imprimir_transicao(de(EstadoOrigem, Simbolo, EstadoDestino)) :-
    format("  ~w --(~w)--> ~w~n", [EstadoOrigem, Simbolo, EstadoDestino]). % Mostra transição no formato origem-símbolo-destino

% Verifica se uma palavra é aceita pelo autômato
teste_palavra(automato(Alfabeto, EstadoInicial, EstadosFinais, Transicoes), Palavra, Resultado) :-
    string_chars(Palavra, Simbolos), % Divide a palavra em lista de símbolos
    (   forall(member(Simbolo, Simbolos), member(Simbolo, Alfabeto)) -> % Confere se os símbolos pertencem ao alfabeto
        simular_transicoes(Transicoes, EstadoInicial, Simbolos, EstadoFinal, EstadosPercorridos), % Executa a simulação
        (   member(EstadoFinal, EstadosFinais) -> % Verifica se o estado final é de aceitação
            Aceita = true
        ;   Aceita = false
        )
    ;   _EstadoFinal = "Fora do alfabeto", % Define mensagem para símbolos inválidos
        EstadosPercorridos = ["Fora do alfabeto"],
        Aceita = false
    ),
    Resultado = resultado_teste{palavra: Palavra, aceita: Aceita, estados_percorridos: EstadosPercorridos}. % Estrutura o resultado

% Simula o processamento da palavra no autômato
simular_transicoes(_, EstadoAtual, [], EstadoAtual, [EstadoAtual]). % Caso base: retorna o estado atual ao fim da palavra
simular_transicoes(Transicoes, EstadoAtual, [Simbolo|Restante], EstadoFinal, [EstadoAtual|EstadosPercorridos]) :-
    member(de(EstadoAtual, Simbolo, ProximoEstado), Transicoes), % Busca transição válida
    simular_transicoes(Transicoes, ProximoEstado, Restante, EstadoFinal, EstadosPercorridos). % Processa o restante da palavra

% Salva os resultados dos testes em um arquivo JSON
salvar_resultados(Arquivo, Resultados) :-
    NomePasta = 'pastaResultados/', % Define o diretório de saída
    (   exists_directory(NomePasta) -> % Verifica existência do diretório
        true
    ;   make_directory(NomePasta), % Cria o diretório se necessário
        writeln('Pasta "pastaResultados" criada.')
    ),
    atom_concat(NomePasta, Arquivo, CaminhoAtom), % Gera o caminho completo do arquivo
    atom_string(CaminhoAtom, CaminhoCompleto), % Converte o caminho para string
    maplist(converter_resultado_para_json, Resultados, ResultadosJSON), % Converte resultados para formato JSON
    open(CaminhoCompleto, write, Stream), % Abre o arquivo em modo escrita
    json_write_dict(Stream, json{resultados: ResultadosJSON}, [pretty(true)]), % Grava os resultados no arquivo
    close(Stream), % Fecha o fluxo de escrita
    format("Resultados salvos em ~w~n", [CaminhoCompleto]). % Confirma a gravação

% Converte um resultado de teste para o formato JSON
converter_resultado_para_json(resultado_teste{palavra: Palavra, aceita: Aceita, estados_percorridos: Estados}, 
                               json{palavra: Palavra, aceita: Aceita, estados_percorridos: Estados}). % Mapeia campos do resultado para JSON